posthog/async_migrations/migrations/0001_events_sample_by.py

```diff
@@ -0,0 +1,163 @@
+            """,
+            rollback=f"""
+                RENAME TABLE
+                    {EVENTS_TABLE_NAME} to {EVENTS_TABLE_NAME}_failed,
+                    {BACKUP_TABLE_NAME} to {EVENTS_TABLE_NAME}
+                ON CLUSTER {CLICKHOUSE_CLUSTER}
+            """,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"OPTIMIZE TABLE {EVENTS_TABLE_NAME} FINAL",
+            rollback="",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", True),
+            side_effect_rollback=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
+        ),
+    ]
+
+    def is_required(self):
+        res = sync_execute(f"SHOW CREATE TABLE {EVENTS_TABLE_NAME}")
+        return "ORDER BY (team_id, toDate(timestamp), cityHash64(distinct_id), cityHash64(uuid))" not in res[0][0]
+
+    def precheck(self):
+        result = sync_execute(
+            f"""
+        SELECT (free_space.size / greatest(event_table_size.size, 1)) FROM 
+            (SELECT 1 as jc, 'event_table_size', sum(bytes) as size FROM system.parts WHERE table = 'sharded_events' AND database='{CLICKHOUSE_DATABASE}') event_table_size
+        JOIN 
+            (SELECT 1 as jc, 'free_disk_space', free_space as size FROM system.disks WHERE name = 'default') free_space
+        ON event_table_size.jc=free_space.jc 
+        """
+        )
+        event_size_to_free_space_ratio = result[0][0]
+
+        # Require 1.5x the events table in free space to be available
+        if event_size_to_free_space_ratio > 1.5:
+            return (True, None)
+        else:
+            result = sync_execute(
+                f"""
+            SELECT formatReadableSize(free_space.size - (free_space.free_space - (1.5 * event_table_size.size ))) as required FROM 
+                (SELECT 1 as jc, 'event_table_size', sum(bytes) as size FROM system.parts WHERE table = 'sharded_events' AND database='{CLICKHOUSE_DATABASE}') event_table_size
+            JOIN 
+                (SELECT 1 as jc, 'free_disk_space', free_space, total_space as size FROM system.disks WHERE name = 'default') free_space
+            ON event_table_size.jc=free_space.jc
+            """
+            )
+            required_space = result[0][0]
+            return (False, f"Upgrade your ClickHouse storage to at least {required_space}.")
+
+    def healthcheck(self):
+        result = sync_execute("SELECT free_space FROM system.disks")
+        # 100mb or less left
+        if int(result[0][0]) < 100000000:
+            return (False, "ClickHouse available storage below 100MB")
+
+        return (True, None)
+
+    def progress(self, _):
```

Comment:
There are other steps in this migration too, a bit weird to see 100% when we're not done yet